lightwood/data/splitter.py

```diff
@@ -1,48 +1,84 @@
+# TODO: Make stratification work for regression via histogram bins??
+from lightwood.api.dtype import dtype
 import pandas as pd
 import numpy as np
-from typing import List
+from typing import List, Dict
 from itertools import product
-
 from lightwood.api.types import TimeseriesSettings
 
 
-def splitter(data: pd.DataFrame, k: int, tss: TimeseriesSettings, seed: int) -> List[pd.DataFrame]:
-    """
-    Splits a dataframe into k equally-sized subsets.
+def splitter(
+    data: pd.DataFrame,
+    tss: TimeseriesSettings,
+    dtype_dict: Dict[str, str],
+    seed: int,
+    pct_train: int,
+    pct_dev: int,
+    pct_test: int,
+    target: str
+) -> Dict[str, pd.DataFrame]:
     """
+    Splits a dataset into stratified training/test. First shuffles the data within the dataframe (via ``df.sample``).
+
+    :param data: Input dataset to be split
+    :param tss: time-series specific details for splitting
+    :param pct_train: training fraction of data; must be less than 1
+    :param dtype_dict: Dictionary with the data type of all columns
+    :param seed: Random state for pandas data-frame shuffling
+    :param n_subsets: Number of subsets to create from data (for time-series)
+    :param target: Name of the target column; if specified, data will be stratified on this column
+
+    :returns: A dictionary containing the keys train, test and dev with their respective data frames, as well as the "stratified_on" key indicating which columns the data was stratified on (None if it wasn't stratified on anything)
+    """ # noqa
+    if pct_train + pct_dev + pct_test != 100:
+        raise Exception('The train, dev and test percentage of the data needs to sum up to 100')
+
+    gcd = np.gcd(100, np.gcd(pct_test, np.gcd(pct_train, pct_dev)))
+    nr_subsets = int(100 / gcd)
+
+    # Shuffle the data
     if not tss.is_timeseries:
-        # shuffle
         data = data.sample(frac=1, random_state=seed).reset_index(drop=True)
 
-        # split
-        subsets = np.array_split(data, k)
-
+    stratify_on = None
+    if tss.is_timeseries or dtype_dict[target] in (dtype.categorical, dtype.binary) and target is not None:
+        stratify_on = [target]
+        if isinstance(tss.group_by, list):
+            stratify_on = stratify_on + tss.group_by
```

Comment:
I believe this will impact the stratification for TS tasks. At the moment it is only done using `tss.group_by`. May be worth trying, but I'd default to doing `stratify_on = tss.group_by`.

Reply:
I assume timeseries tasks with categorical targets are rare, in the cases they exist, it seems reasonable to also stratify on that

Reply:
Fair point, that is nice. However, does this work for numerical TS? Unless I'm reading this wrong, it would still stratify on the target then, which I'm not sure is correct.